### ISSUE TITLE
MinkowskiConvolution with an even-sized kernel and expand_coordinates=True

### ISSUE THREAD

**Issue by peiyunh** [NVIDIA/MinkowskiEngine#328]
**Describe the bug**
MinkowskiConvolution seems to produce incorrect results with an even-sized kernel and `expand_coordinates=True`. 

**To Reproduce**
Below is a piece of code where we perform 2x2 convolution on a 2x2 1-D feature map. 
```Python
import torch
import MinkowskiEngine as ME

data = [
    [0, 2],
    [1, 3]
]

def to_sparse_coo(data):
    coords, feats = [], []
    for i, row in enumerate(data):
        for j, val in enumerate(row):
            if val != 0:
                coords.append([i, j])
                feats.append([val])
    return torch.IntTensor(coords), torch.FloatTensor(feats)

coords, feats = to_sparse_coo(data)
C, F = ME.utils.sparse_collate(coords=[coords], feats=[feats])
sin = ME.SparseTensor(coordinates=C, features=F)
conv = ME.MinkowskiConvolution(in_channels=1,
                               out_channels=1,
                               kernel_size=[2, 2],
                               stride=[1, 1],
                               dimension=2,
                               expand_coordinates=True)
torch.nn.init.constant_(conv.kernel, 1)
sout = conv(sin)
print(sout.dense()[0])
```
On my machine, the code above produces the following output:
```Python
tensor([[[[0., 5., 0.],
          [4., 3., 0.],
          [0., 0., 0.]]]], grad_fn=<CopySlices>)
```

**Expected behavior**
Unless I missed something, shouldn't the output be the following? 
```Python
tensor([[[[6., 5.],
          [4., 3.]]]], grad_fn=<CopySlices>)
```

**Desktop (please complete the following information):**
```
==========System==========
Linux-5.8.0-44-generic-x86_64-with-glibc2.10
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.2 LTS"
3.8.8 (default, Feb 24 2021, 21:46:12) 
[GCC 7.3.0]
==========Pytorch==========
1.7.1
torch.cuda.is_available(): True
==========NVIDIA-SMI==========
/usr/bin/nvidia-smi
Driver Version 455.32.00
CUDA Version 11.1
VBIOS Version 90.02.17.00.5F
Image Version G001.0000.02.04
==========NVCC==========
/usr/local/cuda/bin/nvcc
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Mon_Oct_12_20:09:46_PDT_2020
Cuda compilation tools, release 11.1, V11.1.105
Build cuda_11.1.TC455_06.29190527_0
==========CC==========
/usr/bin/c++
c++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

==========MinkowskiEngine==========
0.5.2
MinkowskiEngine compiled with CUDA Support: True
NVCC version MinkowskiEngine is compiled: 11010
CUDART version MinkowskiEngine is compiled: 11010
```

**Additional context**
Here is another example with a different set of kernel weights.
```Python
import torch
import MinkowskiEngine as ME

data = [
    [0, 2],
    [1, 3]
]

def to_sparse_coo(data):
    coords, feats = [], []
    for i, row in enumerate(data):
        for j, val in enumerate(row):
            if val != 0:
                coords.append([i, j])
                feats.append([val])
    return torch.IntTensor(coords), torch.FloatTensor(feats)

coords, feats = to_sparse_coo(data)
C, F = ME.utils.sparse_collate(coords=[coords], feats=[feats])
sin = ME.SparseTensor(coordinates=C, features=F)
conv = ME.MinkowskiConvolution(in_channels=1,
                               out_channels=1,
                               kernel_size=[2, 2],
                               stride=[1, 1],
                               dimension=2,
                               expand_coordinates=True)
torch.nn.init.constant_(conv.kernel, 0)
conv.kernel[1::2] = 1
sout = conv(sin)
print(sout.dense()[0])
```

On my machine, the code produces the following output
```Python
tensor([[[[0., 3., 0.],
          [0., 0., 0.],
          [0., 0., 0.]]]], grad_fn=<CopySlices>)
```

But shouldn't the correct output be the following:
```Python
tensor([[[[3., 3.],
          [3., 0.]]]], grad_fn=<CopySlices>)
```

**Comment by chrischoy** [comment]
This is the expected behavior.
The input coordinates look like this

```
   o
o o
```

The expanded coordinates which apply 2x2 at every `o` will look like

```
   o o
o o o
o o o
```

So at (0,0) we do not have value and it will be 0.


On a different note,
Also, the output is

```
tensor([[[[0., 5., 0.],
          [4., 3., 0.],
          [0., 0., 0.]]]], grad_fn=<CopySlices>)
```

but if you use ConvolutionTranspose you get

```
tensor([[[[0., 2., 2.],
          [1., 6., 5.],
          [1., 4., 3.]]]], grad_fn=<CopySlices>)
```

**Comment by peiyunh** [comment]
I appreciate the reply. My understanding came from the documentation. Based on what you said, shouldn't the subscript of `x` that is inside the summation be `u-i` instead of `u+i`? 

Documentation URL: https://nvidia.github.io/MinkowskiEngine/convolution.html
![Selection_001](https://user-images.githubusercontent.com/10396605/113795136-30498300-971a-11eb-82c0-5f6d6b1c87b8.png)

**Comment by chrischoy** [comment]
On the bottom note, it says "For even K, ... $\mathbf{i} \in [0, K)^D$"

So, $\{u + i | i \in [0, K)^D\}$ is the location the current kernel would pool features from.

**Comment by peiyunh** [comment]
Thanks for pointing out the note. Yeah, I agree with you that it is `u+i` and not `u-i`. 

But I still think the output produced by the code for the example above is wrong. The output should be
```
A = tensor([[[[0., 2., 2.],
              [1., 6., 5.],
              [1., 4., 3.]]]], grad_fn=<CopySlices>)
```
rather than
```
B = tensor([[[[0., 5., 0.],
              [4., 3., 0.],
              [0., 0., 0.]]]], grad_fn=<CopySlices>)
```

Below is my explanation of why `A` is the correct output. 

First, we have to acknowledge that it is possible for even-sized kernels with `expand_coordinates=True` to produce an output with negative coordinates, because `x[u+i]` could be non-zero even when `u` is negative. 

Of course, from the coding perspective, we don't want negative coordinates. But we can offset them afterward to make sure they start at 0. However, the result *before* such shifting happens should look like: 
```
[0 2 2
 1 6 5 
 1 4 3]
```
with coordinates being
```
[(-1,-1), (-1, 0), (-1, 1), 
 ( 0,-1), ( 0, 0), ( 0, 1), 
 ( 1,-1), ( 1, 0), ( 1, 1)]
```

For a sanity check, I implemented the same toy example in PyTorch and compared the output of PyTorch's `conv2d` to the output of `MinkowskiConvolution`. Assume under the same condition, the results should match, but they don't. Here is the code: 
```
import torch
import MinkowskiEngine as ME

data = [
    [0, 2],
    [1, 3]
]
'''
kernel = [
    [1, 1],
    [1, 1]
]
'''

def to_sparse_coo(data):
    coords, feats = [], []
    for i, row in enumerate(data):
        for j, val in enumerate(row):
            if val != 0:
                coords.append([i, j])
                feats.append([val])
    return torch.IntTensor(coords), torch.FloatTensor(feats)


# ME
coords, feats = to_sparse_coo(data)
C, F = ME.utils.sparse_collate(coords=[coords], feats=[feats])
sin = ME.SparseTensor(coordinates=C, features=F)
me_conv = ME.MinkowskiConvolution(in_channels=1,
                                  out_channels=1,
                                  kernel_size=[2, 2],
                                  stride=[1, 1],
                                  dimension=2,
                                  expand_coordinates=True)
torch.nn.init.constant_(me_conv.kernel, 1)
sout = me_conv(sin)
print(sout.dense()[0])

# torch
torch_conv = torch.nn.Conv2d(in_channels=1,
                             out_channels=1,
                             kernel_size=[2, 2],
                             stride=[1, 1],
                             padding=((1, 1)),
                             bias=False)
torch.nn.init.constant_(torch_conv.weight, 1)

din = torch.FloatTensor(data)

dout = torch_conv(din[None, None, :, :])
print(dout)
```

Here is the output of `MinkowskiConvolution` 
```
tensor([[[[0., 5., 0.],
          [4., 3., 0.],
          [0., 0., 0.]]]], grad_fn=<CopySlices>)
```

And here is the output of PyTorch's `conv2d`
```
tensor([[[[0., 2., 2.],
          [1., 6., 5.],
          [1., 4., 3.]]]], grad_fn=<ThnnConv2DBackward>)
```

I have an initial guess of why the code ends up outputting `B`, but I want to see if you agree with what I wrote above. Did I make sense?

**Comment by chrischoy** [comment]
The MinkowskiEngine SparseTensor supports negative coordinates and you do not need to translate or pad the input.

First, the expand coordinate expands using the kernel size==2. Thus, the expanded coordinates look like

  o o
o o o
o o o


Next, for the coordinate (0,2), the convolution equation will take four coordinates as inputs: (0,2),(0,3),(1,2),(1,3). However, features at these coordinates are 0s. Thus, the output is 0.

If you use expand coordinate with kernels size 3 first and then convolve with kernel size 2, you would get the same output as the torch conv2d with padding.

**Comment by peiyunh** [comment]
Thanks for the response. By coordinate (0, 2), I assume you mean the coordinate of `[o]`
```
  o [o]
o o  o
o o  o
```

This is where my understanding differs. To me, the coordinate of `[o]` is (-1, 1) instead of (0, 2). Therefore, the convolution equation takes these four coordinates as input: (-1, 1), (-1, 2), (0, 1), (0, 2). The features at these coordinates are 0, 0, 2, 0. Thus, the output is 2. Am I wrong?

**Comment by chrischoy** [comment]
`expand_coordinate`  use the same offset ${u + i | i \in [0, K)^D}$ (when K is even) to create new coordinates.
For example, kernel_size=2 will expand a coordinate (0, 1) to (0, 1), (0, 2), (1, 1), (1, 2). So in the example you mentioned, `[o]` is (0,2) not (-1, 1).

If you use kernel_size=3, it will use ${u + i | i \in [-(K-1)/2, (K - 1)/2]^D}$ where K=3.

**Comment by peiyunh** [comment]
Thanks for the explanation. I now see where my understanding differs. 

What I thought `expand_coordinate` does is computing 
<img src="https://latex.codecogs.com/gif.latex?\inline&space;\large&space;\mathcal{C}^\text{out}&space;=&space;\{\mathbf{u}&space;|&space;\mathbf{u}&plus;\mathbf{i}&space;\in&space;\mathcal{C}^\text{in}&space;\text{~if&space;any~}&space;\mathbf{i}&space;\in&space;[0,&space;K)^D&space;\}" title="\large \mathcal{C}^\text{out} = \{\mathbf{u} | \mathbf{u}+\mathbf{i} \in \mathcal{C}^\text{in} \text{~if any~} \mathbf{i} \in [0, K)^D \}" />

Plus, if we were to follow the convolution equation and compute the output (-1,1), we would have
<img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;\mathbf{x}_{\mathbf{u}=(-1,1)}&space;&=&space;\sum_{\mathbf{i}&space;\in&space;\{0,1\}^2}&space;W_\mathbf{i}&space;\mathbf{x}_{\mathbf{u}&space;&plus;&space;\mathbf{i}}\\&space;&=&space;W_{(0,0)}&space;\mathbf{x}_{(-1,1)}&space;&plus;&space;W_{(0,1)}&space;\mathbf{x}_{(-1,2)}&space;&plus;&space;W_{(1,0)}&space;\mathbf{x}_{(0,1)}&space;&plus;&space;W_{(1,1)}&space;\mathbf{x}_{(0,2)}\\&space;&=0&space;&plus;&space;0&space;&plus;&space;2&space;&plus;&space;0\\&space;&=2&space;\end{align*}" title="\begin{align*} \mathbf{x}_{\mathbf{u}=(-1,1)} &= \sum_{\mathbf{i} \in \{0,1\}^2} W_\mathbf{i} \mathbf{x}_{\mathbf{u} + \mathbf{i}}\\ &= W_{(0,0)} \mathbf{x}_{(-1,1)} + W_{(0,1)} \mathbf{x}_{(-1,2)} + W_{(1,0)} \mathbf{x}_{(0,1)} + W_{(1,1)} \mathbf{x}_{(0,2)}\\ &=0 + 0 + 2 + 0\\ &=2 \end{align*}" />

Do you spot anything wrong in my thought process?

**Comment by chrischoy** [comment]
The out coordinates are defined the same way the corresponding convolution kernel region is defined: $\{u + i| u \in \mathcal{C}_{in}, i \in \{0, 1\}^D\}$ so for (0,1), it would be (0,1), (0,2), (1,1), (1,2).

**Comment by peiyunh** [comment]
Got you. Thanks for the replies!
